### PR TITLE
Signed score diffs

### LINKED ISSUE #10: Compute Score Diff Instead of Abs. Score Diff
Currently in `src.aggregations.primitive_pairs`, we are computing the absolute value of the score diff between each pair of pipelines. If we don't take the absolute value, we have more information about which primitive's tend to perform better than others. The only complication in this feature request is that we need the order of the calculation of score difference to be consistent in regards to the pair of primitives.

### PR DESCRIPTION
**Pull request by JeremyRees**
Resolves #10. The issue itself is solved in the first commit, [Don't take absolute value of score diffs](https://github.com/byu-dml/d3m-mtl-db-reader/commit/dffeb4680a88f1017c7e36c8cfb4afb27f32e6b7).

The rest of the pull request is the creation of unit tests to check the correct ordering of everything in the primitive pair aggregation. Along with this I added tests for pipeline run entities and other tests for the primitive pair aggregation. Because pipeline runs reference pipelines, datasets, problems, and predictions, I had to create a lot of new test data, which makes up the bulk of this PR.

The commit [Allow specifying run predictions path](https://github.com/byu-dml/d3m-mtl-db-reader/commit/c06aa156ca5803e43bf4fdc900ac40925410f55e) was necessary because previously the `PipelineRun` entity was hard-coded to find predictions under the `DATA_ROOT` directory, but we needed to change that for our unit tests. To maintain consistency, `**kwargs` was also added to the constructor for `Dataset`s, `Pipeline`s, and `Problem`s.

### CODE REVIEW

File: src/aggregations/primitive_pairs.py [hunk 1]
```diff
@@ -12,8 +12,14 @@
 
 
 class ScoreDiff(NamedTuple):
+    """
+    `metric_score_diff` is calculated as the the score of `run_a` (in the
+    containing PipelineRunPairDiffEntry) minus the score of `run_b`. I.e.,
+    a positive value indicates that `run_a` had a higher score, while a
+    negative value means `run_b` had a higher score.
+    """
```
Comment (epeters3) [inline src/aggregations/primitive_pairs.py:20]: Beautiful documentation.

Comment (epeters3) [reply src/aggregations/primitive_pairs.py:20]: It looks like down below on line 181 the actual computation is `run_b_score.value - run_a_score.value`.

File: src/entities/pipeline_run.py [hunk 2]
```diff
@@ -50,22 +50,26 @@ def __init__(self, pipeline_run_dict: dict):
...
-        self.init_predictions(pipeline_run_dict)
+        self.init_predictions(pipeline_run_dict, **kwargs)
 
         # These references will be dereferenced later by the loader
         # once the pipelines, problems, and datasets are available.
         self.datasets: List[Union[DocumentReference, Dataset]] = []
         for dataset_dict in pipeline_run_dict["datasets"]:
             self.datasets.append(DocumentReference(dataset_dict))
-        self.pipeline: Union[DocumentReference, Problem] = DocumentReference(
+        self.pipeline: Union[DocumentReference, Pipeline] = DocumentReference(
```
Comment (epeters3) [inline src/entities/pipeline_run.py:63]: Good catch.

File: src/entities/pipeline_run.py [hunk 3]
```diff
@@ -50,22 +50,26 @@ def __init__(self, pipeline_run_dict: dict):
...
             pipeline_run_dict["pipeline"]
         )
         self.problem: Union[DocumentReference, Problem] = DocumentReference(
             pipeline_run_dict["problem"]
         )
 
-    def init_predictions(self, pipeline_run_dict) -> None:
+    def init_predictions(self, pipeline_run_dict, **kwargs) -> None:
```
Comment (epeters3) [inline src/entities/pipeline_run.py]: I think the more common convention is to document all used parameters in the function signature. Maybe we could change the function signature to something like `def init_predictions(self, pipeline_run_dict, *, run_predictions_path = DataDir.PREDICTIONS_DUMP.value, **kwargs) -> None:`

Comment (epeters3) [reply src/entities/pipeline_run.py]: Then we wouldn't need the `kwargs.get(...)` call.

File: tests/utils.py [hunk 4]
```diff
@@ -1,32 +1,95 @@
...
+        'type': PipelineRun,
+        'names': [
+            'similar_pipeline_run_a',
+            'similar_pipeline_run_b',
+            'similar_pipeline_run_c',
+            'similar_pipeline_run_d',
+            'invalid_pipeline_run_a', # Has a state of FAILURE
+            'invalid_pipeline_run_b', # Has no predictions
+            'invalid_pipeline_run_c', # Phase is FIT, not PRODUCE
+            'invalid_pipeline_run_d', # Problem task type is unsupported
```
Comment (epeters3) [inline tests/utils.py:36]: This documentation of what each test data document contains and can be used for is great. Would you be willing to add a small comment like this to the similar pipeline runs and invalid problem as well?

**Review by epeters3 — CHANGES_REQUESTED** [review]
This is beautiful Jeremy. I appreciate you describing the test data you added, and the unit tests are so nice to have. Please see my comments on this PR.

**Review by epeters3 — CHANGES_REQUESTED** [review]
This is beautiful Jeremy. I appreciate you describing the test data you added, and the unit tests are so nice to have. Please see my comments on this PR.

File: src/aggregations/primitive_pairs.py (hunk 1 above)
Comment (JeremyRees) [reply src/aggregations/primitive_pairs.py:20]: Oops, good catch!

**Review by epeters3 — APPROVED** [review]

**Pull request merged**